crates/core/src/metadata_store.rs

```diff
@@ -465,75 +542,26 @@ impl From<WriteError> for ReadWriteError {
                 unreachable!("failed preconditions should be treated separately")
             }
             WriteError::Network(err) => ReadWriteError::Network(err),
-            WriteError::Internal(msg) => ReadWriteError::Internal(msg),
+            WriteError::Other(err) => ReadWriteError::Other(err),
             WriteError::Codec(err) => ReadWriteError::Codec(err),
-            WriteError::Store(err) => ReadWriteError::Store(err),
-        }
-    }
-}
-
-pub trait MetadataStoreClientError {
-    fn is_network_error(&self) -> bool;
-}
-
-impl<E> MetadataStoreClientError for ReadModifyWriteError<E> {
-    fn is_network_error(&self) -> bool {
-        match self {
-            ReadModifyWriteError::ReadWrite(err) => err.is_network_error(),
-            ReadModifyWriteError::FailedOperation(_) => false,
-        }
-    }
-}
-
-impl MetadataStoreClientError for ReadWriteError {
-    fn is_network_error(&self) -> bool {
-        // we don't use the catch all pattern _ to make sure
-        // new errors are explicitly handled
-        match self {
-            ReadWriteError::Network(_) => true,
-            ReadWriteError::Internal(_) => false,
-            ReadWriteError::Codec(_) => false,
-            ReadWriteError::RetriesExhausted(_) => false,
-            ReadWriteError::Store(_) => false,
-        }
-    }
-}
-
-impl MetadataStoreClientError for WriteError {
-    fn is_network_error(&self) -> bool {
-        match self {
-            WriteError::Network(_) => true,
-            WriteError::FailedPrecondition(_) => false,
-            WriteError::Internal(_) => false,
-            WriteError::Codec(_) => false,
-            WriteError::Store(_) => false,
-        }
-    }
-}
-
-impl MetadataStoreClientError for ReadError {
-    fn is_network_error(&self) -> bool {
-        match self {
-            ReadError::Network(_) => true,
-            ReadError::Codec(_) | ReadError::Internal(_) | ReadError::Store(_) => false,
         }
     }
 }
 
 static_assertions::assert_impl_all!(MetadataStoreClient: Send, Sync, Clone);
 
-pub async fn retry_on_network_error<Fn, Fut, T, E, P>(retry_policy: P, action: Fn) -> Result<T, E>
+pub async fn retry_on_retryable_error<Fn, Fut, T, E, P>(retry_policy: P, action: Fn) -> Result<T, E>
 where
     P: Into<RetryPolicy>,
     Fn: FnMut() -> Fut,
     Fut: Future<Output = Result<T, E>>,
-    E: MetadataStoreClientError + std::fmt::Display,
+    E: MaybeRetryableError,
 {
     retry_policy
         .into()
         .retry_if(action, |err: &E| {
-            if err.is_network_error() {
-                trace!("could not connect to metadata store: {err}; retrying");
+            if err.retryable() {
+                trace!("operation failed: {err}; retrying");
```

Comment:
Probably debug level? we'd want to see if a metadata operation is being retried, right?

Reply:
Good point. Will change it.